src/bin/pg_autoctl/nodestate_utils.c

```diff
@@ -0,0 +1,306 @@
+ * characteristics.
+ */
+void
+nodestateAdjustHeaders(NodeAddressHeaders *headers,
+					   NodeAddress *node, int groupId)
+{
+	int nameLen = strlen(node->name);
+
+	/* compute strlen of host:port */
+	IntString portString = intToString(node->port);
+	int hostLen =
+		strlen(node->host) + strlen(portString.strValue) + 1;
+
+	/* compute strlen of groupId/nodeId, as in "0/1" */
+	IntString nodeIdString = intToString(node->nodeId);
+	int nodeLen = 0;
+
+	switch (headers->nodeKind)
+	{
+		case NODE_KIND_STANDALONE:
+		{
+			nodeLen = strlen(nodeIdString.strValue);
+			break;
+		}
+
+		default:
+		{
+			IntString groupIdString = intToString(groupId);
+
+			nodeLen =
+				strlen(groupIdString.strValue) +
+				strlen(nodeIdString.strValue) + 1;
+
+			break;
+		}
+	}
+
+	/* initialize to mininum values, if needed */
+	if (headers->maxNameSize == 0)
+	{
+		/* Name */
+		headers->maxNameSize = 5;
+	}
+
+	if (headers->maxHostSize == 0)
+	{
+		/* Host:Port */
+		headers->maxHostSize = 10;
+	}
+
+	if (headers->maxNodeSize == 0)
+	{
+		/* groupId/nodeId */
+		headers->maxNodeSize = 5;
+	}
+
+	if (nameLen > headers->maxNameSize)
+	{
+		headers->maxNameSize = nameLen;
+	}
+
+	if (hostLen > headers->maxHostSize)
+	{
+		headers->maxHostSize = hostLen;
+	}
+
+	if (nodeLen > headers->maxNodeSize)
+	{
+		headers->maxNodeSize = nodeLen;
+	}
+}
+
+
+/*
+ * nodestatePrintHeader prints the given CurrentNodeStateArray header.
+ */
+void
+nodestatePrintHeader(NodeAddressHeaders *headers)
+{
+	fformat(stdout, "%*s | %*s | %*s | %17s | %17s | %17s | %6s\n",
+			headers->maxNameSize, "Name",
+			headers->maxNodeSize, "Node",
+			headers->maxHostSize, "Host:Port",
+			"Current State", "Assigned State",
+			"LSN", "Health");
+
+	fformat(stdout, "%*s-+-%*s-+-%*s-+-%17s-+-%17s-+-%17s-+-%6s\n",
+			headers->maxNameSize,
+			headers->nameSeparatorHeader,
+			headers->maxNodeSize,
+			headers->nodeSeparatorHeader,
+			headers->maxHostSize,
+			headers->hostSeparatorHeader,
+			"-----------------", "-----------------", "-----------------",
+			"------");
+}
+
+
+/*
+ * nodestatePrintNodeState prints the node at the given position in the given
+ * nodesArray, using the nodesArray pre-computed sizes for the dynamic columns.
+ */
+void
+nodestatePrintNodeState(NodeAddressHeaders *headers,
+						CurrentNodeState *nodeState)
+{
+	char hostport[BUFSIZE] = { 0 };
+	char composedId[BUFSIZE] = { 0 };
+
+	(void) nodestatePrepareNode(headers,
+								&(nodeState->node),
+								nodeState->groupId,
+								hostport,
+								composedId);
+
+	fformat(stdout, "%*s | %*s | %*s | %17s | %17s | %17s | %6s\n",
+			headers->maxNameSize, nodeState->node.name,
+			headers->maxNodeSize, composedId,
+			headers->maxHostSize, hostport,
+			NodeStateToString(nodeState->reportedState),
+			NodeStateToString(nodeState->goalState),
+			nodeState->node.lsn,
+			nodeState->health == -1 ? "?" :
+			nodeState->health == 1 ? "✓" : "✗");
```

Comment:
Maybe instead of `Health` this should be called `Reachable` or something similar. It wasn't obvious to me that this `Health` didn't include anything about the current state, but simply if the monitor could reach the node. I'm not sure if it still makes sense then to put it in front of `Current State`.

Reply:
In terms of user experience it's not the best, because of unicode character display width. Even with a monospace font, chars such as ✓ and ✗ will take a different width when printed on the terminal. This will then break our attempt at having any alignment at all in the output for most users. That's the reason why this is the only column with unicode, and it has to be the last field on the output.

Reply:
What do you think of renaming he column to `Reachable` instead of `Healthy` for clarity?

Also, we could simply write them out in words as well. `Yes`, `No` and `?` instead of using unicode. If that's causing issues.

Reply:
Renamed to Reachable, that's a good point, and then used the strings "unknown", "yes", and "no". Makes it easier.

Reply:
And now moved the Reachable field before the State field, with the headers being all "computed" it's easy now.